data/transactions/logic/eval.go

```diff
@@ -2571,6 +2571,76 @@ func opSetByte(cx *evalContext) {
 	cx.stack = cx.stack[:prev]
 }
 
+func extract(x []byte, start, length int) (out []byte, err error) {
+	out = x
+	end := start + length
+	if start > len(x) || end > len(x) {
+		err = errors.New("extract range beyond length of string")
+		return
+	}
+	out = x[start:end]
+	err = nil
+	return
+}
+
+func opExtract(cx *evalContext) {
+	last := len(cx.stack) - 1
+	start := cx.program[cx.pc+1]
+	length := cx.program[cx.pc+2]
+	// Shortcut: if length is 0, take bytes from start index to the end
+	l := int(length)
+	if length == 0 {
+		l = len(cx.stack[last].Bytes) - int(start)
+	}
+	cx.stack[last].Bytes, cx.err = extract(cx.stack[last].Bytes, int(start), l)
+}
+
+func opExtract3(cx *evalContext) {
+	last := len(cx.stack) - 1 // length
+	prev := last - 1          // start
+	pprev := prev - 1         // bytes
+	start := cx.stack[prev].Uint
+	length := cx.stack[last].Uint
+	if start > math.MaxInt32 || length > math.MaxInt32 {
+		cx.err = errors.New("extract range beyond length of string")
+		return
+	}
+	cx.stack[pprev].Bytes, cx.err = extract(cx.stack[pprev].Bytes, int(start), int(length))
+	cx.stack = cx.stack[:prev]
+}
+
+func convertBytesToInt(x []byte) (out uint64) {
```

Comment:
I think this is the same as the `binary.BigEndian.Uint64()` function from the `encoding/binary` package?

Reply:
You're right, it looks quite similar -- I just lifted this from the [`btoi` function](https://github.com/algorand/go-algorand/blame/master/data/transactions/logic/eval.go#L1080)...

Reply:
Oh interesting. I'm not familiar enough with the binary syntax to know whether they are equivalent; I wonder what @jannotti thinks.

Reply:
I wondered the same, whether we should be using a library function.  But when I saw the btoi function did it this way, I didn't bother saying anything.  It would be great to investigate and see if we can use the library. Always preferable not to do things like this ourselves.  One thing to confirm - the library call need to accept "short" byte arrays. A single byte is a legal uint64 decoded "our way".

Reply:
No dice.
```
func (bigEndian) Uint64(b []byte) uint64 {
	_ = b[7] // bounds check hint to compiler; see golang.org/issue/14808
	return uint64(b[7]) | uint64(b[6])<<8 | uint64(b[5])<<16 | uint64(b[4])<<24 |
		uint64(b[3])<<32 | uint64(b[2])<<40 | uint64(b[1])<<48 | uint64(b[0])<<56
}
```

Reply:
Good catch JJ -- I'll leave it as is then.

Reply:
You could extract it into a function if you'd like.

Either way, could you add a comment explaining why we do it?

Reply:
Done!